src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -0,0 +1,253 @@
+				(SymbolKind.TypeParameter, SymbolKind.Method) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsMethodReturnType,
+				(SymbolKind.TypeParameter, SymbolKind.Parameter) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsParameter,
+				(SymbolKind.TypeParameter, SymbolKind.TypeParameter) => DiagnosticId.DynamicallyAccessedMembersMismatchTypeArgumentTargetsGenericParameter,
+				_ => throw new NotImplementedException ()
+			};
+
+		static string[] GetDiagnosticArguments (ISymbol source, ISymbol target, string missingAnnotations)
+		{
+			var args = new List<string> ();
+			args.AddRange (GetDiagnosticArguments (target));
+			args.AddRange (GetDiagnosticArguments (source));
+			args.Add (missingAnnotations);
+			return args.ToArray ();
+		}
+
+		static IEnumerable<string> GetDiagnosticArguments (ISymbol symbol)
+		{
+			var args = new List<string> ();
+			args.AddRange (symbol.Kind switch {
+				SymbolKind.Parameter => new string[] { symbol.GetDisplayName (), symbol.ContainingSymbol.GetDisplayName () },
+				SymbolKind.NamedType => new string[] { symbol.GetDisplayName () },
+				SymbolKind.Field => new string[] { symbol.GetDisplayName () },
+				SymbolKind.Method => new string[] { symbol.GetDisplayName () },
+				SymbolKind.TypeParameter => new string[] { symbol.GetDisplayName (), symbol.ContainingSymbol.GetDisplayName () },
+				_ => throw new NotImplementedException ($"Unsupported source or target symbol {symbol}.")
+			});
+
+			return args;
+		}
+
+		static ISymbol? GetSymbolFromOperation (IOperation? operation) =>
+			operation switch {
+				IArgumentOperation argument => GetSymbolFromOperation (argument.Value),
+				IAssignmentOperation assignment => GetSymbolFromOperation (assignment.Value),
+				IConversionOperation conversion => conversion.Type,
+				IInstanceReferenceOperation instanceReference => instanceReference.Type,
+				IInvocationOperation invocation => invocation.TargetMethod,
+				IMemberReferenceOperation memberReference => memberReference.Member,
+				IParameterReferenceOperation parameterReference => parameterReference.Parameter,
+				IReturnOperation returnOp => GetSymbolFromOperation (returnOp.ReturnedValue),
+				ITypeOfOperation typeOf => typeOf.TypeOperand,
+				_ => null
+			};
+
+		static bool SourceHasMatchingAnnotations (
+			DynamicallyAccessedMemberTypes? sourceMemberTypes,
+			DynamicallyAccessedMemberTypes? targetMemberTypes,
+			out string missingMemberTypesString)
+		{
+			missingMemberTypesString = $"'{nameof (DynamicallyAccessedMemberTypes.All)}'";
+			if (targetMemberTypes == null)
+				return true;
+
+			sourceMemberTypes ??= DynamicallyAccessedMemberTypes.None;
+			var missingMemberTypesList = Enum.GetValues (typeof (DynamicallyAccessedMemberTypes))
+				.Cast<DynamicallyAccessedMemberTypes> ()
+				.Where (damt => (damt & targetMemberTypes & ~sourceMemberTypes) == damt && damt != DynamicallyAccessedMemberTypes.None)
```

Comment:
Would it be possible to extract this logic from `RequireDynamicallyAccessedMembers` and share it between the linker and analyzer?

Reply:
Good point. We could use the same logic here and in the ReflectionMethodBodyScanner.